specs/contract_signatures/01_main.rst

```diff
@@ -0,0 +1,78 @@
+==========================================================
+Contract Signatures
+==========================================================
+
+=====  ===================
+epic      title       
+=====  ===================
+`78`_  Contract Signatures
+=====  ===================
+
+.. _78: https://github.com/gnosis/safe/issues/78
+
+.. _Main:
+
+
+#. `Main`_
+
+.. sectnum::
+.. contents:: Table of Contents
+    :local:
+    :depth: 2
+
+Problem Definition
+---------------------
+
+
+Technical Details
+-----------------
+
+Diagram_
+
+1. Dapp sends a JSON payload request to the browser extension.
+
+2. The chome extension signs the hash of the payload according to EIP-712_ and sends the following payload to the app (via push notification)
+
+.. code:: javascript
+    
+    {
+        "devices": ["<checksumed_address>", ...],
+        "message": "<string>",
+        "signature": { // signs sha3("GNO" + <message>)
+            "v": "<integer>",
+            "r": "<string>", // stringified int
+            "s": "<string>" // stringified int
+        }
+    }
+
+::
+
+Where ``message`` is the following:
+
+.. code:: javascript
+    
+    {
+        "type": "signTypedData"
+        "payload": <string> // Payload to sign
+        "signed-hash": <hex-string> // Signature of the hash of the payload according to EIP712
+        "address": <checksummed_address> // Address of the signer (Browser Extension)
+    }
+
+
+3. App: recover the address from the payload and verifies that it matches the address from the browser extension.
+4. App: signs the payload and waits for the other signatures from other possible signers (until threshold is reached).
+5. App sends push to the chrome extension with all the signatures. The ``message`` of the push notification is the following:
+
+.. code:: javascript
+    
+    {
+        "type": "signTypedDataResult"
+        "payload": <string> // Payload to sign
+        "signed-hashes": [<hex-string>, ...] // Signature of the hash of the payload according to EIP712

```

Comment:
I would return a `signature` field that is a `<hex-string>`  that is already the final encoding

Reply:
Do you mean like:

```json
{
    "signed-hashes": [<hex-string>, ...],
    "signatures": [<r,s,v object>, ...]
}
```
?

Forgot the signatures for final verification

Reply:
I would not return arrays ... all clients should sign the same hash. and I would already encode the signatures according to the safe docs (https://gnosis-safe.readthedocs.io/en/latest/contracts/signatures.html)